main.py

```diff
@@ -1,11 +1,126 @@
+import time
+from io import BytesIO
+from PIL import Image
 import logger
 import sound as snd
+import pygame.image
+import camera
 import tcp
-
+import udp
+import struct
 from threading import Thread
 
 buffer_size = None
+INFORMATION_SIZE = struct.calcsize("i" * 3)
+CHUNK_SIZE = 1500 - INFORMATION_SIZE
+
+
+def get_and_send_data(sock, resolution=(640, 480)):
+    sock.settimeout(100)
+    cam = camera.camera_init()
+
+    current_image_number = 0
+    try:
+        while True:
+            while True:
+                if cam.query_image():
+                    camera_image = cam.get_image()
+                    break
+
+            buffer = BytesIO()
+            im = Image.frombuffer("RGB", resolution, bytes(pygame.image.tostring(camera_image, "RGB")))
+            # NOTE: quality parameter was chosen experimentally
+            im.save(buffer, optimize=True, quality=45, format='JPEG')
+            image = buffer.getvalue()
+            buffer.close()
+
+            # To prevent integer overflow in bytes sending
+            if current_image_number == 2 ** 32 - 1:
+                current_image_number = 0
+
+            for i in range(len(image) // CHUNK_SIZE + (len(image) / CHUNK_SIZE != 0)):
+                sock.send(b''.join([struct.pack('i', len(image)), struct.pack('i', current_image_number),
+                                    struct.pack('i', i), image[i * CHUNK_SIZE:(i + 1) * CHUNK_SIZE]]))
+
+            current_image_number += 1
+            logger.root_logger.debug(f"Camera. Got and sent {len(image)} bytes")
+            time.sleep(0)
+
+    except (BrokenPipeError, ConnectionResetError) as e:
+        logger.root_logger.warning(e)
+
+
+def receive_data(sock, queue=b'', resolution=(640, 480)):
+    sock.settimeout(100)
+    log = logging.getLogger("Camera")
+    window_display = pygame.display.set_mode(resolution)
+
+    current_image_number = 0
+    try:
+        while True:
+            if len(queue) != 0:
+                data = [queue]
+            else:
+                data = []
+
+            # To prevent integer overflow in bytes sending
+            if current_image_number == 2 ** 32 - 1:
+                current_image_number = 0
+
+            while True:
+                try:
+                    pack = sock.recv(CHUNK_SIZE + INFORMATION_SIZE)
+                    while struct.unpack('i', pack[4:8])[0] < current_image_number:
+                        pack = sock.recv(CHUNK_SIZE + INFORMATION_SIZE)
+                        pass
+                    if struct.unpack('i', pack[4:8])[0] == current_image_number:
+                        data += [pack]
+                    if struct.unpack('i', pack[4:8])[0] > current_image_number:
+                        queue = pack
+                        break
+                except TimeoutError:
+                    pass
+
+            current_image_number += 1
+
+            data = list(set(data))
+            data.sort(key=lambda data_item: struct.unpack('i', data_item[8:12])[0])
```

Comment:
Just look at this line. Isn't it beautiful? I hope you love it as much as I do.

Anyway, all this sorting business can be avoided as you have indices of all those elements. Just assign them to array using these indices and that's all.

P.S. yes, first line is pure sarcasm :ok_man: